Add Klarna to the available payment methods

a-danae opened the pull request:
Fixes #3043 

## Changes proposed in this Pull Request:

- Accept payments with Klarna using the Updated checkout experience. 

## Testing instructions

Find the testing instructions in [v8.3.0 ‐ Release testing instructions](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/v8.3.0-%E2%80%90-Release-testing-instructions#klarna)

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [x] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

a-danae committed:
Display Klarna on the payment methods settings tab

a-danae committed:
Adjust php unit tests

a-danae committed:
Add Klarna icon on the shortcode checkout page

a-danae committed:
Add domestic payment restrictions and multiple currency support for Klarna

a-danae committed:
Merge branch 'develop' into add/klarna-payment-method

a-danae committed:
Fix error when the account doesn't have a default currency

a-danae committed:
Set an instance of WC_Stripe_Account as a property of WC_Stripe_UPE_Payment_Method

a-danae committed:
Adjust failing tests due to the introduced accept_only_domestic_payment check

a-danae committed:
Inject WC_Stripe_Account as a dependency of WC_Stripe_UPE_Payment_Gateway

a-danae committed:
Adjust unit tests

a-danae committed:
Remove currency in the Klarna payment method description

a-danae committed:
Add base 'supported_countries' property to WC_Stripe_UPE_Payment_Method

a-danae committed:
Restrict payment methods on the block checkout based on the customer billing country

a-danae committed:
Hide payment methods on the shortcode checkout based on the customer billing country

a-danae committed:
Merge branch 'develop' into add/klarna-payment-method

a-danae committed:
Update unit tests for the payment fields html output

a-danae committed:
Keep  as the first argument of the is_enabled_at_checkout method

a-danae committed:
Add php unit tests for has_domestic_transactions_restrictions

a-danae committed:
Remove unit test leftover from a previous implementation

james-allan commented on client/stripe-utils/utils.js:
```diff
@@ -489,3 +489,41 @@ export const initializeUPEAppearance = ( api, isBlockCheckout = 'false' ) => {
 };
+
+/**
+ * Hides payment method if it has set specific countries in the PHP class.
```
```suggestion
 * Determines if the payment method is restricted to specific countries. 
```

I note that this is a copy of WooPayments function but just a minor suggestion. This function description suggests that it will hide the payment element when in fact it just returns a bool that can be used for that purpose.

a-danae replied:
True, thanks! 🙂 Suggestion committed.

james-allan commented on client/stripe-utils/utils.js:
```diff
@@ -489,3 +489,41 @@ export const initializeUPEAppearance = ( api, isBlockCheckout = 'false' ) => {
+	// in the case of "pay for order", there is no "billing country" input, so we need to rely on backend data.
+	const billingCountry =
+		document.getElementById( 'billing_country' )?.value ||
+		wcpayCustomerData?.billing_country ||
```
This `wcpayCustomerData` won't exist in the Stripe context. I think it needs to be `getStripeServerData()`

<img width="652" alt="Screenshot 2024-04-24 at 2 15 09 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/e1b2fb61-371a-4061-99a7-aaca25323ff1">

james-allan replied:
Looking into this a little more, the `getStripeServerData()` which comes from the `wc_stripe_upe_params` localized params which are generated in [this  `javascript_params()` function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/klarna-payment-method/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L355) doesn't include the billing country.

We might need to do something similar to this: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/add/klarna-payment-method/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L2002-L2016

a-danae replied:
Whoops. Great catch! Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3040/commits/09dd5b8b78c3103e92308369e005d0a95858966a

james-allan replied:
Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3040/commits/09dd5b8b78c3103e92308369e005d0a95858966a

james-allan reviewed (changes requested):
Hey @a-danae. This PR is looking pretty good. I left 2 initial review comments which are a minor comment and fixing an error.

I ran through the following tests. Noting that I tested with an account based in the US. I also ran some basic tests on a AU account just to make sure things work there too. 

**Checkout**
- [x] Not shown for subscription purchases. 
- [x] Not shown for purchases not in invalid currency.
- [x] Not shown when billing address is in invalid country.
- [x] Not shown when auth and capture are enabled.
- [x] ⚠️ Purchase outside the merchants base location (ie based in the US, customer in AU). 
     -  There's an error that occurs on the checkout but it seems irrelevant to the actual underlying issue. See errors below.

| Checkout Error    | Stripe Error |
| -------- | ------- |
| <img width="1161" alt="Screenshot 2024-04-24 at 3 10 05 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/8cbde4ba-368f-4ab6-97a8-1cfbfba38e65"> | <img width="600" alt="Screenshot 2024-04-24 at 3 11 45 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/224dc85b-2697-46e2-97d7-70a38456325f">  |

- [x] Completed purchase using 4 instalments. 
- [x] Abandon payment (don't confirm after checkout submit).
    - There's an error notice shown on the shortcode checkout indicating payment couldn't be completed. There's no notice on the block checkout though I think it's primed and refreshing the page will make it show. (see below). 
- [x] Failed payment using denied phone number. [ref](https://docs.stripe.com/payments/klarna/accept-a-payment?web-or-mobile=web&payments-ui-type=direct-api&country=united-states#testmode-guide)
 
<img width="600" alt="Screenshot 2024-04-24 at 3 19 58 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/634557e2-1859-4c97-9e37-d302d9f1733d">

**Pay for Order**
- [x] Not shown for purchases not in invalid currency.
- [x] Completed purchase. 

**My Account**
- [x] Not shown on payment methods page. 
- [x] Cannot add saved token. 

**Admin**
- [x] Order correctly records Klarna as the Payment method.
- [x] Full Refunds
- [x] Partial Refunds

**Stripe Dashboard**
- [x] Payments correctly reflect Klarna as the Payment method.
- [x] Refunds

**Testing instructions**

- [x] Double check the tests I've ran cover the official testing instructions in [v8.3.0 ‐ Release testing instructions](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/v8.3.0-%E2%80%90-Release-testing-instructions#klarna)

#### A comment on designs

This is a minor thing and I'd suggest we handle them separately as to not unnecessarily hold up this PR, but in the designs for WooPayments that we're basing our on, I noticed their Klarna icon in the admin section and on the block checkout is filled, whereas ours has a white section

| Designs    | Current |
| -------- | ------- |
| <img width="300" alt="Screenshot 2024-04-24 at 2 47 39 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/bb5e1b05-8bbe-478d-bb41-1dadc7b6a917"><img width="150" alt="Screenshot 2024-04-24 at 2 48 17 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/ada07edf-dba3-419e-a58f-fca6535f6f9c"> | <img width="300" alt="Screenshot 2024-04-24 at 4 53 26 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/54eeb990-35a6-402e-bb61-bd3701e88b3c"><img width="150" alt="Screenshot 2024-04-24 at 4 52 51 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/424a9270-e661-4e16-a927-89f6bc99af02"> |

I'll continue testing on Friday as I'll be afk tomorrow. :)

a-danae committed:
Update function description

This function description suggests that it will hide the payment element when in fact it just returns a bool that can be used for that purpose.

Co-authored-by: James Allan <james.allan@automattic.com>

a-danae committed:
Pass the billing country to the localized JS params on the Pay for order page

a-danae commented:
Thanks for the thorough review so far, @james-allan ! 🙂 

I've created a couple of issues for these behaviors:

> ⚠️ Purchase outside the merchants base location (ie based in the US, customer in AU).
There's an error that occurs on the checkout but it seems irrelevant to the actual underlying issue. See errors below.

This is also happening in WooPayments btw. I opened an issue over there too.
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3096

> in the designs for WooPayments that we're basing our on, I noticed their Klarna icon in the admin section and on the block checkout is filled, whereas ours has a white section

Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3040/commits/daba441e4f8c4044ff21251ced614bd2f1fcc3c2. It'd be good to unify how we're displaying all of the icons, but it's not a priority for now.

~Not blocking this PR makes sense to me! Here's the related issue~
- ~https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3097~

a-danae committed:
Fix Klarna icon sizing

james-allan commented:
Changes look good. I've finished all my testing instructions and reviewed the list you mentioned in the 8.3 list. :)

a-danae committed:
Merge branch 'develop' into add/klarna-payment-method

a-danae committed:
Add changelog entries
